File: multiqc/plots/echarts/box.py
"""
ECharts option builder for `PlotType.BOX` (`multiqc/plots/box.py`).

MultiQC box plots are horizontal: samples on the (inverted) category axis (y), values
on the value axis (x), mirroring bar.py. See `templates/echarts/src/js/plots/box.js` for
the interactive (toolbox-aware) JS counterpart of `series()`/`axis_data()` below.

ECharts' native `boxplot` series type takes a PRECOMPUTED five-number summary
(`[min, q1, median, q3, max]`) per box; unlike Plotly's `go.Box`, it never computes
quartiles from raw points itself. Outlier points (values outside the box whiskers) are
drawn as a companion `scatter` series, the same convention ECharts' own
`echarts.dataTool.prepareBoxplotData` helper uses.

Quartile method (risk flagged in `multiqc-echarts-exploration/BUILD_PLAN.md` section 4,
"Box quartile formulas"): linear interpolation between the two closest ranks, i.e.
numpy's default `np.percentile` method and R's type-7 quantile, which is also Plotly's
default `quartilemethod="linear"` for `go.Box`. Whisker ends are the most extreme
in-fence data points (Tukey convention, 1.5 * IQR), matching both Plotly's
`lowerfence`/`upperfence` and ECharts' own boxplot data-prep tool; so `min`/`max` in the
five-number summary below are NOT the sample's absolute min/max, but the whisker ends;
values outside the fence are outliers.

This is the SAME formula ported verbatim to JS in
`templates/echarts/src/js/plots/box.js` (`quantile`/`fiveNumberSummary`/`outliers`); the
cross-language contract is pinned by one golden fixed-input/expected-output test in
`tests/test_plots_echarts.py` (`test_box_five_number_summary_golden_values` /
`test_box_outliers_golden_values`), whose exact input and expected values are mirrored
in a comment block at the top of that JS file.

For `is_stats_data` datasets (`Dataset` in `multiqc/plots/box.py`, pre-calculated
per-sample statistics, used for memory efficiency on huge datasets), the five-number
summary is read directly from the sample's `min`/`q1`/`median`/`q3`/`max` keys: we do
not (and, lacking raw values, cannot) recompute quartiles or outliers for these samples.
"""

from typing import TYPE_CHECKING, Any, Dict, List, Tuple, Union, cast

from multiqc.plots.box import BoxPlotConfig, BoxStatsT, Dataset
from multiqc.plots.echarts.converter import convert_layout

if TYPE_CHECKING:
    from multiqc.plots.plot import Plot


def _quantile(sorted_values: List[float], q: float) -> float:
    """
    Linear-interpolation quantile (numpy's default `np.percentile` method, R's type-7,
    and Plotly's default `quartilemethod="linear"`). `sorted_values` must already be
    sorted ascending and non-empty.
    """
    n = len(sorted_values)
    if n == 1:
        return sorted_values[0]
    h = (n - 1) * q
    lo = int(h)  # floor: h is always >= 0
    hi = min(lo + 1, n - 1)
    frac = h - lo
    return sorted_values[lo] + frac * (sorted_values[hi] - sorted_values[lo])


def _fences(q1: float, q3: float) -> Tuple[float, float]:
    iqr = q3 - q1
    return q1 - 1.5 * iqr, q3 + 1.5 * iqr


def five_number_summary(values: List[Union[int, float]]) -> List[float]:
    """
    `[min, q1, median, q3, max]` for one sample's raw data points, in the format the
    ECharts `boxplot` series expects. `min`/`max` here are the boxplot WHISKER ends (the
    most extreme data points within 1.5 * IQR of the box, Tukey convention), not the
    sample's absolute min/max; see the module docstring.
    """
    sorted_values = sorted(float(v) for v in values)
    q1 = _quantile(sorted_values, 0.25)
    median = _quantile(sorted_values, 0.5)
    q3 = _quantile(sorted_values, 0.75)
    lower_fence, upper_fence = _fences(q1, q3)
    in_fence = [v for v in sorted_values if lower_fence <= v <= upper_fence]
    whisker_min = in_fence[0] if in_fence else sorted_values[0]
    whisker_max = in_fence[-1] if in_fence else sorted_values[-1]
    return [whisker_min, q1, median, q3, whisker_max]


def outliers(values: List[Union[int, float]]) -> List[float]:
    """Values outside the Tukey 1.5 * IQR fence: the points the whisker box doesn't reach."""
    sorted_values = sorted(float(v) for v in values)
    q1 = _quantile(sorted_values, 0.25)
    q3 = _quantile(sorted_values, 0.75)
    lower_fence, upper_fence = _fences(q1, q3)
    return [v for v in sorted_values if v < lower_fence or v > upper_fence]


def _stats_five_number(stats: BoxStatsT) -> List[float]:
    return [float(stats["min"]), float(stats["q1"]), float(stats["median"]), float(stats["q3"]), float(stats["max"])]


def layout_option(plot: "Plot[Any, Any]", dataset: Dataset) -> Dict[str, Any]:
    """
    Full ECharts option skeleton for one box-plot dataset, minus `series` and axis
    `data` arrays (samples are toolbox-dependent; the JS renderer fills `yAxis.data`).
    """
    # xAxis is the value axis (box plots are horizontal); Plotly-style data-fitted
    # autorange instead of ECharts' forced-0 default, see `converter._convert_axis`.
    option = convert_layout(plot.layout, dataset.layout, scale_x=True)

    option["yAxis"]["type"] = "category"
    option["yAxis"]["inverse"] = True
    option["tooltip"]["trigger"] = "item"

    return option


def _build(dataset: Dataset) -> Tuple[List[List[float]], List[Dict[str, Any]]]:
    """
    Shared computation for `series()` and `mark_count()`, so the two can never disagree
    on how many marks are actually drawn: the boxplot five-number data (one entry per
    sample, in `dataset.samples` order) and the companion outlier/point scatter data.
    """
    boxpoints = dataset.trace_params.get("boxpoints", "outliers")

    box_data: List[List[float]] = []
    scatter_data: List[Dict[str, Any]] = []
    for yi, (sample, values) in enumerate(zip(dataset.samples, dataset.data)):
        if dataset.is_stats_data:
            box_data.append(_stats_five_number(cast(BoxStatsT, values)))
            continue

        raw_values = cast(List[Union[int, float]], values)
        if not raw_values:
            box_data.append([0.0, 0.0, 0.0, 0.0, 0.0])
            continue
        box_data.append(five_number_summary(raw_values))

        if boxpoints is False:
            continue
        shown = sorted(float(v) for v in raw_values) if boxpoints == "all" else outliers(raw_values)
        scatter_data.extend({"name": sample, "value": [value, yi]} for value in shown)

    return box_data, scatter_data


def series(dataset: Dataset, pconfig: BoxPlotConfig, is_pct: bool) -> List[Dict[str, Any]]:
    """
    One `{"type": "boxplot"}` series (one five-number entry per sample, positionally
    matching `yAxis.data`) plus, unless `boxpoints` is `False`, a companion
    `{"type": "scatter"}` series of outlier points (or every point, when
    `boxpoints == "all"`). This is the SSR/get_option (non-toolbox) path; the
    interactive path is `EchartsBoxPlot.buildSeries()`
    (`templates/echarts/src/js/plots/box.js`).

    `pconfig`/`is_pct` are accepted for dispatch-signature parity with `bar.series`; box
    plots never enable the percentage switch, so they are unused here.
    """
    box_data, scatter_data = _build(dataset)
    result: List[Dict[str, Any]] = [{"type": "boxplot", "data": box_data}]
    if scatter_data:
        result.append({"type": "scatter", "name": "outliers", "data": scatter_data, "symbolSize": 5})
    return result


def axis_data(dataset: Dataset, pconfig: BoxPlotConfig) -> List[Tuple[str, List[str]]]:
    """`[("yAxis", sample names)]`: box plots are always horizontal, samples on `yAxis`."""
    return [("yAxis", list(dataset.samples))]


def mark_count(dataset: Dataset) -> int:
    """One mark per sample (box), plus one per outlier/shown point."""
    box_data, scatter_data = _build(dataset)
    return len(box_data) + len(scatter_data)

File: multiqc/templates/default/src/js/plots-echarts/box.js
// ECharts box plot: extends the default template's BoxPlot (imported just before this
// file in main-js.js) for its constructor/prepData()/exportData()/
// formatDatasetForAiPrompt()/activeDatasetSize(), which are engine-neutral: prepData()
// already resolves the sort-switch (dataset.data_sorted/samples_sorted) and applies
// toolbox hide/rename via applyToolboxSettings(). window.BoxPlot itself extends
// window.Plot, which by this point in the import order is OUR echarts Plot class (from
// echarts-plotting.js), not Plotly's. The default box.js sort-toggle click handler at
// its bottom (`renderPlot(anchor)`) is engine-neutral and reused as-is.
//
// GOLDEN quartile fixture (cross-language contract with
// multiqc/plots/echarts/box.py, asserted in
// tests/test_plots_echarts.py::test_box_five_number_summary_golden_values /
// test_box_outliers_golden_values):
//   input:            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 100]
//   fiveNumberSummary: [1, 3.5, 6, 8.5, 10]
//   outliers:          [100]
// fiveNumberSummary()/outliers() below MUST stay byte-for-byte the same formula as the
// Python functions of the same name in multiqc/plots/echarts/box.py.

// Linear-interpolation quantile (numpy's default `np.percentile` method, R's type-7,
// and Plotly's default `quartilemethod="linear"`); mirrors
// multiqc/plots/echarts/box.py::_quantile. `sortedValues` must already be sorted
// ascending and non-empty.
function quantile(sortedValues, q) {
  const n = sortedValues.length;
  if (n === 1) return sortedValues[0];
  const h = (n - 1) * q;
  const lo = Math.floor(h);
  const hi = Math.min(lo + 1, n - 1);
  const frac = h - lo;
  return sortedValues[lo] + frac * (sortedValues[hi] - sortedValues[lo]);
}

// [min, q1, median, q3, max] as ECharts' boxplot series expects, mirroring
// multiqc/plots/echarts/box.py::five_number_summary. min/max here are the boxplot
// WHISKER ends (most extreme in-fence points, Tukey 1.5*IQR convention), not the
// sample's absolute min/max; values outside are outliers (see outliers() below).
function fiveNumberSummary(values) {
  let sorted = [...values].sort((a, b) => a - b);
  let q1 = quantile(sorted, 0.25);
  let median = quantile(sorted, 0.5);
  let q3 = quantile(sorted, 0.75);
  let iqr = q3 - q1;
  let lowerFence = q1 - 1.5 * iqr;
  let upperFence = q3 + 1.5 * iqr;
  let inFence = sorted.filter((v) => v >= lowerFence && v <= upperFence);
  let whiskerMin = inFence.length ? inFence[0] : sorted[0];
  let whiskerMax = inFence.length ? inFence[inFence.length - 1] : sorted[sorted.length - 1];
  return [whiskerMin, q1, median, q3, whiskerMax];
}

// Values outside the Tukey 1.5*IQR fence, mirroring multiqc/plots/echarts/box.py::outliers.
function outliers(values) {
  let sorted = [...values].sort((a, b) => a - b);
  let q1 = quantile(sorted, 0.25);
  let q3 = quantile(sorted, 0.75);
  let iqr = q3 - q1;
  let lowerFence = q1 - 1.5 * iqr;
  let upperFence = q3 + 1.5 * iqr;
  return sorted.filter((v) => v < lowerFence || v > upperFence);
}

// [r, g, b] from any CSS color (hex, named like "grey", or an existing rgb()/rgba()
// string). Hex/named colors are resolved via a throwaway canvas context, which always
// normalizes an opaque color to "#rrggbb".
function colorToRgbTriplet(color) {
  let match = color.match(/rgba?\(([^)]+)\)/);
  if (match)
    return match[1]
      .split(",")
      .slice(0, 3)
      .map((s) => s.trim());
  let hex = Object.assign(document.createElement("canvas").getContext("2d"), { fillStyle: color }).fillStyle;
  return [1, 3, 5].map((i) => parseInt(hex.slice(i, i + 2), 16));
}

// Match Plotly's own box rendering exactly (verified against the default template's
// output: a single stroked SVG path per box at stroke-opacity 1 for the border,
// whiskers AND median line, with fill-opacity 0.5 for the body). ECharts draws the
// median line using itemStyle.borderColor, so a solid border against a lighter,
// semi-transparent fill is what makes the median visible instead of blending into a
// fully opaque box. Mirrors multiqc/plots/echarts/box.py::_box_item_style.
function boxItemStyle(color) {
  let [r, g, b] = colorToRgbTriplet(color);
  return { color: `rgba(${r},${g},${b},0.5)`, borderColor: `rgb(${r},${g},${b})` };
}

class EchartsBoxPlot extends window.BoxPlot {
  // Constructs and returns the ECharts boxplot series (+ companion outlier scatter
  // series) for the active dataset, applying the sort switch and toolbox
  // (hide/rename/highlight) via the inherited prepData().
  buildSeries() {
    let [data, samples] = this.prepData();
    if (data.length === 0 || samples.length === 0) return [];

    let dataset = this.datasets[this.activeDatasetIdx];
    let traceParams = dataset["trace_params"] || {};
    let boxpoints = "boxpoints" in traceParams ? traceParams["boxpoints"] : "outliers";
    let baseColor = (traceParams["marker"] && traceParams["marker"]["color"]) || "#4899e8";

    // axisLabel (item B): colors the sample-name axis tick labels, mirroring Plotly's
    // recalculateTicks(); see Plot.sampleAxisLabel() in echarts-plotting.js. maxTicks
    // mirrors the Plotly box plot's own formula (templates/plotly/src/js/plots/box.js).
    let maxTicks = (this.layout.height - 140) / 12;
    this._axisData = {
      axis: "yAxis",
      data: this.filteredSettings.map((s) => s.name),
      axisLabel: this.sampleAxisLabel(this.filteredSettings, maxTicks),
    };

    let boxData = [];
    let scatterData = [];

    this.filteredSettings.forEach((sample, i) => {
      // Same highlight rule as default box.js's Plotly buildTraces(): recolor the
      // highlighted sample to its highlight color, every other sample to grey, once
      // ANY highlight is active anywhere (item A: global flag, not the local
      // `highlighted.length` count, so this plot greys out fully even with zero matches).
      let color = window.mqc_highlight_f_texts.length > 0 ? (sample.highlight ?? "grey") : baseColor;
      let values = data[i];

      if (this.isStatsData) {
        boxData.push({
          value: [values.min ?? 0, values.q1 ?? 0, values.median ?? 0, values.q3 ?? 0, values.max ?? 0],
          itemStyle: boxItemStyle(color),
        });
        return;
      }

      boxData.push({ value: fiveNumberSummary(values), itemStyle: boxItemStyle(color) });

      if (boxpoints === false) return;
      let shown = boxpoints === "all" ? [...values].sort((a, b) => a - b) : outliers(values);
      shown.forEach((value) => {
        scatterData.push({ name: sample.name, value: [value, i], itemStyle: { color: color } });
      });
    });

    let series = [{ type: "boxplot", layout: "horizontal", data: boxData }];
    if (scatterData.length > 0) {
      series.push({ type: "scatter", name: "outliers", data: scatterData, symbolSize: 5 });
    }
    return series;
  }

  // Tooltip formatter can't live in the JSON-safe serialized skeleton, so it's attached
  // here to the fully-assembled option (see Plot.applyOptionOverrides in
  // echarts-plotting.js). Shows the full five-number summary for a box, or the single
  // value for an outlier point.
  applyOptionOverrides(option) {
    if (!option.tooltip) return;
    let suffix = this.layout?.xaxis?.ticksuffix ?? "";
    // Box's value axis is xAxis (boxes are horizontal, samples on the category yAxis),
    // matching the Plotly reference (bar.js's own `layout.xaxis?.hoverformat` read applies
    // equally here since box.py also lays out horizontally). NOTE: read off the active
    // dataset's raw layout dict, not `this.layout` (see plots-echarts/bar.js's matching
    // comment): the neutral `AxisIR` behind `this.layout` drops `hoverformat`, a
    // Plotly-only key.
    let hoverformat = this.datasets[this.activeDatasetIdx]?.layout?.xaxis?.hoverformat;
    option.tooltip.formatter = (params) => {
      if (params.seriesType === "boxplot") {
        // ECharts prepends the category axis value/index to `value` for a boxplot on a
        // category axis (6 elements: [axisValue, min, q1, median, q3, max]), unlike the
        // plain 5-element five-number array in option.series[0].data; drop it here.
        let [minV, q1, median, q3, maxV] = params.value
          .slice(-5)
          .map((v) => window.formatWithHoverformat(v, hoverformat));
        return (
          `<b>${params.name}</b><br/>` +
          `Max: ${maxV}${suffix}<br/>Q3: ${q3}${suffix}<br/>Median: ${median}${suffix}<br/>` +
          `Q1: ${q1}${suffix}<br/>Min: ${minV}${suffix}`
        );
      }
      let value = Array.isArray(params.value) ? params.value[0] : params.value;
      return `<b>${params.name}</b>: ${window.formatWithHoverformat(value, hoverformat)}${suffix}`;
    };
  }
}

// Make EchartsBoxPlot globally available (referenced by bare window.EchartsBoxPlot in
// echarts-plotting.js's initPlot(), not a bare identifier, so it's robust across the
// minified bundle regardless of module ordering after bundling).
window.EchartsBoxPlot = EchartsBoxPlot;
